# Salary Prediction System using Linear Regression
## Objective
The objective of this project is to build a Machine Learning model that predicts employee salaries based on years of experience using the Linear Regression algorithm.

This project demonstrates the complete Machine Learning workflow including:

- Loading a dataset
- Data cleaning and preprocessing
- Handling missing values
- Splitting data into training and testing sets
- Training a Linear Regression model
- Making predictions
- Evaluating model performance
- Saving the trained model using Pickle and Joblib

# Import Required Libraries

The following libraries are used:

- Pandas → Data manipulation
- NumPy → Numerical operations
- Scikit-Learn → Machine Learning algorithms
- Pickle → Save trained models
- Joblib → Efficient model storage

In [2]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

import pickle
import joblib

# Load Dataset

The dataset is loaded using Pandas.
The dataset contains employee information such as:
- Age
- Gender
- Education Level
- Job Title
- Years of Experience
- Salary

In [3]:
df = pd.read_csv("Salary_Data.csv")

df.head()

,Age,Gender,Education Level,Job Title,Years of Experience,Salary
0,32.0,Male,Bachelor's,Software Engineer,5.0,90000.0
1,28.0,Female,Master's,Data Analyst,3.0,65000.0
2,45.0,Male,PhD,Senior Manager,15.0,150000.0
3,36.0,Female,Bachelor's,Sales Associate,7.0,60000.0
4,52.0,Male,Master's,Director,20.0,200000.0


# Explore Dataset

Before training a model, it is important to understand:
- Number of rows and columns
- Data types
- Missing values
- Statistical summary

In [4]:
print(df.shape)

df.info()

df.describe()

(6704, 6)
<class 'pandas.DataFrame'>
RangeIndex: 6704 entries, 0 to 6703
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Age                  6702 non-null   float64
 1   Gender               6702 non-null   str    
 2   Education Level      6701 non-null   str    
 3   Job Title            6702 non-null   str    
 4   Years of Experience  6701 non-null   float64
 5   Salary               6699 non-null   float64
dtypes: float64(3), str(3)
memory usage: 314.4 KB


,Age,Years of Experience,Salary
count,6702.000000,6701.000000,6699.000000
mean,33.620859,8.094687,115326.964771
std,7.614633,6.059003,52786.183911
min,21.000000,0.000000,350.000000
25%,28.000000,3.000000,70000.000000
50%,32.000000,7.000000,115000.000000
75%,38.000000,12.000000,160000.000000
max,62.000000,34.000000,250000.000000


# Handling Missing Values

Missing values can negatively affect model performance.
We first identify missing values and then remove them.

In [5]:
df.isnull().sum()

Age                    2
Gender                 2
Education Level        3
Job Title              2
Years of Experience    3
Salary                 5
dtype: int64

# Data Cleaning

Rows containing missing values are removed to improve data quality.
Data cleaning is an important preprocessing step in Machine Learning.

In [6]:
df = df.dropna()

print("Dataset Shape After Cleaning:")
print(df.shape)

Dataset Shape After Cleaning:
(6698, 6)


# Feature Selection
For simplicity, we use:

Feature:(X)
- Years of Experience

Target:(Y)
- Salary
The model will learn how salary changes with experience.

In [7]:
X = df[['Years of Experience']]

y = df['Salary']

# Split Dataset

The dataset is divided into:

- 80% Training Data
- 20% Testing Data

The training set is used to train the model.

The testing set is used to evaluate the model.

In [8]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42
)

print("Training Samples:", len(X_train))
print("Testing Samples:", len(X_test))

Training Samples: 5358
Testing Samples: 1340


# Linear Regression Model

Linear Regression attempts to find the best-fit line between input features and target values.

Formula:
Salary = m × Experience + c

Where:

- m = Slope         X=features (Years of experience)
- c = Intercept     Y=Target  (Salary)

In [9]:
model = LinearRegression()

model.fit(X_train, y_train)

print("Model Trained Successfully")

Model Trained Successfully


# Model Prediction

After training, the model can predict salaries for unseen data.

In [15]:
predictions = model.predict(X_test)

predictions[:10]

array([121851.9658296 ,  86487.65739606,  65269.07233593,  79414.79570935,
        93560.51908277, 121851.9658296 , 100633.38076947,  86487.65739606,
       135997.68920302, 121851.9658296 ])

# Predict Salary for New Employee

The trained model can estimate salary based on years of experience.

In [20]:
import pandas as pd

experience = pd.DataFrame({
    'Years of Experience': [7]
})

predicted_salary = model.predict(experience)

print("Predicted Salary:", predicted_salary[0])

Predicted Salary: 107706.24245618308


# Evaluation Metrics

The model is evaluated using:
## Mean Squared Error (MSE)
Measures average squared prediction error.

Lower value is better.

## Root Mean Squared Error (RMSE)
Square root of MSE.
Represents prediction error in actual salary units.

Lower value is better.

## R² Score
Measures how well the model explains the data.

Range:
- 1.0 → Perfect Model
- 0.8+ → Good Model
- 0.5+ → Acceptable Model

In [21]:
mse = mean_squared_error(y_test, predictions)

rmse = np.sqrt(mse)

r2 = r2_score(y_test, predictions)

print("MSE :", mse)
print("RMSE:", rmse)
print("R² Score:", r2)

MSE : 982050061.9154329
RMSE: 31337.677991763092
R² Score: 0.6555973709206211


# Save Model using Pickle

Pickle allows us to save the trained model and use it later without retraining.

In [22]:
with open("salary_model.pkl", "wb") as file:
    pickle.dump(model, file)

print("Model Saved Successfully")

Model Saved Successfully


# Results

The Linear Regression model was trained using the Salary Dataset.

## Dataset Information

- Total Records: 6704
- Records After Cleaning: 6698
- Missing Values Removed: 6

## Train-Test Split

- Training Data: 80%
- Testing Data: 20%

## Model Performance

- Mean Squared Error (MSE): 982050061.92
- Root Mean Squared Error (RMSE): 31337.68
- R² Score: 0.656

## Interpretation

The model achieved an R² score of 0.656, which means it explains approximately 65.6% of the variation in employee salaries based on years of experience.

The RMSE value indicates that the model's predictions differ from the actual salaries by an average of approximately ₹31,338.

The model provides reasonable predictions but could be improved by including additional features such as education level, age, and job title.